jdtls.ext/com.microsoft.java.maven.plugin/src/main/java/com/microsoft/java/maven/AddDependencyHandler.java

```diff
@@ -0,0 +1,266 @@
+                    textEdits.put(pomUriString, new ArrayList<TextEdit>());
+                } else if (posInfo.alreadyHasDependencies == false) {
+                    final int space = 2;
+                    newtext = linesep + StringUtils.repeat(" ", space) + "<dependencies>" + linesep + 
+                            StringUtils.repeat(" ", space + 2) + "<dependency>" + linesep + 
+                            StringUtils.repeat(" ", space + 4) + "<groupId>" + info[0] + "</groupId>" + 
+                            linesep + StringUtils.repeat(" ", space + 4) + "<artifactId>" + info[1] + 
+                            "</artifactId>" + linesep + StringUtils.repeat(" ", space + 4) + "<version>" + 
+                            info[2] + "</version>" + linesep + StringUtils.repeat(" ", space + 2) + "</dependency>" + 
+                            linesep + StringUtils.repeat(" ", space) + "</dependencies>" + linesep;
+                    textEdits.put(pomUriString, 
+                        Arrays.asList(new TextEdit(new Range(posInfo.pos, posInfo.pos), newtext)));
+                } else {
+                    final int space = posInfo.pos.getCharacter();
+                    newtext = linesep + StringUtils.repeat(" ", space + 2) + "<dependency>" + linesep + 
+                        StringUtils.repeat(" ", space + 4) + "<groupId>" + info[0] + "</groupId>" + linesep + 
+                        StringUtils.repeat(" ", space + 4) + "<artifactId>" + info[1] + "</artifactId>" + 
+                        linesep + StringUtils.repeat(" ", space + 4) + "<version>" + info[2] + "</version>" + 
+                        linesep + StringUtils.repeat(" ", space + 2) + "</dependency>" + linesep + 
+                        StringUtils.repeat(" ", space);
+                    textEdits.put(pomUriString, 
+                        Arrays.asList(new TextEdit(new Range(posInfo.pos, posInfo.pos), newtext)));
+                }
+                return new WorkspaceEdit(textEdits);
+            } else {
+                return new WorkspaceEdit();
+            }
+        } catch (JavaModelException e) {
+            return new WorkspaceEdit();
+        }
+    }
+
+    public static List<WorkspaceEdit> addDependency(AddDependencyParams params, IProgressMonitor monitor) {
+        final ICompilationUnit unit = JDTUtils.resolveCompilationUnit(params.uri);
+        final CompilationUnit astRoot = ASTResolving.createQuickFixAST(unit, null);
+        return Arrays.asList(replaceEdit(params, unit), importEdit(params, unit, astRoot), 
+                            pomEdit(params, unit));
+    }
+}
+
+class GetPosHandler extends DefaultHandler {
+    private Locator locator;
+    private PosInfo posInfo = new PosInfo();
+    private final int lineOffset = 1;
+    private final int columnOffset = 1;
+    private final int dependenciesLength = 15; // the length of "</dependencies>"
+    private final int projectsLength = 10; // the length of "</projects>"
```

Comment:
See naming convention for constants: https://www.oracle.com/technetwork/java/javase/documentation/codeconventions-135099.html